react/features/video-layout/subscriber.js

```diff
@@ -1,24 +1,105 @@
 // @flow
-
 import {
     VIDEO_QUALITY_LEVELS,
     setMaxReceiverVideoQuality
 } from '../base/conference';
-import { StateListenerRegistry } from '../base/redux';
+import {
+    getPinnedParticipant,
+    pinParticipant
+} from '../base/participants';
+import { StateListenerRegistry, equals } from '../base/redux';
 import { selectParticipant } from '../large-video';
 import { shouldDisplayTileView } from './functions';
+import { setParticipantsWithScreenShare } from './actions';
+
+declare var interfaceConfig: Object;
 
 /**
  * StateListenerRegistry provides a reliable way of detecting changes to
  * preferred layout state and dispatching additional actions.
  */
 StateListenerRegistry.register(
     /* selector */ state => shouldDisplayTileView(state),
-    /* listener */ (displayTileView, { dispatch }) => {
+    /* listener */ (displayTileView, store) => {
+        const { dispatch } = store;
+
         dispatch(selectParticipant());
 
         if (!displayTileView) {
-            dispatch(setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+            dispatch(
+                setMaxReceiverVideoQuality(VIDEO_QUALITY_LEVELS.HIGH));
+
+            _updateAutoPinnedParticipant(store);
+        }
+    }
+);
+
+/**
+ * For auto-pin mode, listen for changes to the known media tracks and look
+ * for updates to screen shares.
+ */
+StateListenerRegistry.register(
+    /* selector */ state => state['features/base/tracks'],
+    /* listener */ (tracks, store) => {
+        if (typeof interfaceConfig !== 'object'
+            && !interfaceConfig.AUTO_PIN_LATEST_SCREEN_SHARE) {
+            return;
+        }
+
+        const oldSharingParticipantIds
+            = store.getState()['features/video-layout'].screenShares || [];
+        let newSharingParticipantIds = Array.from(oldSharingParticipantIds);
```

Comment:
Since the `screenShares` is `Array` now why do we still need `Array.from`?

Reply:
To make a copy to modify. I can probably do the filter first and use it’s copy.

Reply:
Why do you need to make a copy? Isn't the `getState` method anyway return a copy of the whole state?

Reply:
Oh you need to compare them later...